src/datastructure/RailwayNetwork.cpp

```diff
@@ -1,655 +1,1610 @@
-    vertex_name_to_index[name] = vertices.size() - 1;
-    return vertex_name_to_index[name];
+  }
+  throw exceptions::EdgeNotExistentException(source_id, target_id);
 }
 
-int cda_rail::Network::add_edge(int source, int target, double length, double max_speed,
-                                 bool breakable, double min_block_length) {
-    /**
-     * Add edge to network
-     * @param source Source vertex
-     * @param target Target vertex
-     * @param length Length of edge
-     * @param max_speed Maximum speed on edge
-     * @param breakable Whether edge is breakable
-     * @param min_block_length Minimum block length on edge
-     */
-    if (source == target) {
-        throw std::invalid_argument("Source and target are the same");
-    }
-    if (!has_vertex(source)) {
-        throw std::out_of_range("Source vertex does not exist");
-    }
-    if (!has_vertex(target)) {
-        throw std::out_of_range("Target vertex does not exist");
+size_t cda_rail::Network::get_edge_index(size_t source_id,
+                                         size_t target_id) const {
+  /**
+   * Get edge index by source and target index
+   *
+   * @param source_id Index of source vertex
+   * @param target_id Index of target vertex
+   *
+   * @return Index of edge
+   */
+  if (!has_vertex(source_id)) {
+    throw exceptions::VertexNotExistentException(source_id);
+  }
+  if (!has_vertex(target_id)) {
+    throw exceptions::VertexNotExistentException(target_id);
+  }
+  for (size_t i = 0; i < edges.size(); i++) {
+    if (edges[i].source == source_id && edges[i].target == target_id) {
+      return i;
     }
-    if (has_edge(source, target)) {
-        throw std::invalid_argument("Edge already exists");
-    }
-    edges.emplace_back(source, target, length, max_speed, breakable, min_block_length);
-    successors.emplace_back();
-    return edges.size() - 1;
+  }
+  throw exceptions::EdgeNotExistentException(source_id, target_id);
 }
 
-void cda_rail::Network::add_successor(int edge_in, int edge_out) {
-    /**
-     * Add successor to edge, but only if the edges are adjacent to each other
-     *
-     * @param edge_in Edge to add successor to
-     * @param edge_out Edge to add as successor
-     */
-    if (!has_edge(edge_in)) {
-        throw std::out_of_range("Edge in does not exist");
-    }
-    if (!has_edge(edge_out)) {
-        throw std::out_of_range("Edge out does not exist");
-    }
-    if (edges[edge_in].target != edges[edge_out].source) {
-        throw std::invalid_argument("Edges are not adjacent");
-    }
+bool cda_rail::Network::has_edge(size_t source_id, size_t target_id) const {
+  /**
+   * Check if edge exists by source and target index
+   *
+   * @param source_id Index of source vertex
+   * @param target_id Index of target vertex
+   *
+   * @return True if edge exists, false otherwise
+   */
+  if (!has_vertex(source_id)) {
+    throw exceptions::VertexNotExistentException(source_id);
```

Comment:
Technically we could also return false here instead of throwing an exception. But I guess it doesn't hurt.

Reply:
I think it makes sense to throw an exception since this is a paraneter that should never be passed in this place.